File: core/src/galenscovell/hinterstar/map/SystemMarker.scala
package galenscovell.hinterstar.map

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d._
import com.badlogic.gdx.scenes.scene2d._
import com.badlogic.gdx.scenes.scene2d.utils.ActorGestureListener
import galenscovell.hinterstar.util._


/**
  * A SystemMarker is a node on the map grid.
  * SystemMarkers are the graphical components of Systems which have sprites and are rendered.
  * SystemMarker also carry state (eg EMPTY, CURRENT, EXPLORED, UNEXPLORED).
  */
class SystemMarker(x: Int, y: Int) extends Actor {
  val sx: Int = x
  val sy: Int = y

  private var frames: Int = 60
  private var markerType: Short = 0
  private var sprite: Sprite = _
  private var glowing: Boolean = true

  this.addListener(new ActorGestureListener() {
    override def touchDown(event: InputEvent, x: Float, y: Float, pointer: Int, button: Int): Unit = {
      if (!isEmpty) {
        SystemRepo.setSelection(getThisSystemMarker)
      }
    }
  })


  /**
    * Return this SystemMarker.
    */
  def getThisSystemMarker: SystemMarker = {
    this
  }

  /**
    * Return true if this SystemMarker is currently EMPTY (no sprite).
    */
  def isEmpty: Boolean = {
    markerType == Constants.SYSTEMMARKER_EMPTY
  }

  /**
    * Return true if this SystemMarker is the CURRENT place the Player is.
    */
  def isCurrent: Boolean = {
    markerType == Constants.SYSTEMMARKER_CURRENT
  }

  /**
    * Return true if this SystemMarker has been EXPLORED by the Player.
    */
  def isExplored: Boolean = {
    markerType == Constants.SYSTEMMARKER_EXPLORED
  }

  /**
    * Return true if this SystemMarker is UNEXPLORED by the Player.
    */
  def isUnexplored: Boolean = {
    markerType == Constants.SYSTEMMARKER_UNEXPLORED
  }



  /**
    * Make this SystemMarker EMPTY (remove its sprite)
    */
  def becomeEmpty(): Unit = {
    markerType = Constants.SYSTEMMARKER_EMPTY
  }

  /**
    * Make this SystemMarker the CURRENT Player System.
    */
  def becomeCurrent(): Unit = {
    Resources.currentMarker.setTarget(
      sx * Constants.SYSTEMMARKER_SIZE - Constants.SYSTEMMARKER_SIZE,
      Gdx.graphics.getHeight - (sy * Constants.SYSTEMMARKER_SIZE) - (6 * Constants.SYSTEMMARKER_SIZE)
    )
    markerType = Constants.SYSTEMMARKER_CURRENT
  }

  /**
    * Mark this SystemMarker as EXPLORED by the Player.
    */
  def becomeExplored(): Unit = {
    markerType = Constants.SYSTEMMARKER_EXPLORED
  }

  /**
    * Mark this SystemMarker as UNEXPLORED by the Player.
    */
  def becomeUnexplored(): Unit = {
    sprite = Resources.spTest0
    markerType = Constants.SYSTEMMARKER_UNEXPLORED
  }



  /**
    * Draw this SystemMarker's sprite.
    */
  override def draw(batch: Batch, parentAlpha: Float): Unit = {
    if (!isEmpty) {
      glow(batch)
      batch.draw(
        sprite, sx * Constants.SYSTEMMARKER_SIZE,
        Gdx.graphics.getHeight - (sy * Constants.SYSTEMMARKER_SIZE) - (5 * Constants.SYSTEMMARKER_SIZE),
        Constants.SYSTEMMARKER_SIZE,
        Constants.SYSTEMMARKER_SIZE
      )
      if (isCurrent) {
        Resources.currentMarker.render(batch)
      }
    }
  }

  /**
    * Draw this SystemMarker's animation effects (glowing/color).
    */
  private def glow(batch: Batch): Unit = {
    if (glowing) {
      frames += 1
    } else {
      frames -= 2
    }

    if (frames == 120) {
      glowing = false
    } else if (frames == 40) {
      glowing = true
    }

    val frameAlpha: Float = frames / 120.0f
    if (isExplored) {
      batch.setColor(0.2f, 0.2f, 0.75f, frameAlpha)
    } else {
      batch.setColor(0.2f, 0.75f, 0.2f, frameAlpha)
    }

    batch.draw(
      Resources.mapGlow,
      sx * Constants.SYSTEMMARKER_SIZE - Constants.SYSTEMMARKER_SIZE,
      Gdx.graphics.getHeight - (sy * Constants.SYSTEMMARKER_SIZE) - (6 * Constants.SYSTEMMARKER_SIZE),
      Constants.SYSTEMMARKER_SIZE * 3,
      Constants.SYSTEMMARKER_SIZE * 3
    )
    batch.setColor(1, 1, 1, 1)
  }
}

File: core/src/galenscovell/hinterstar/generation/sector/SystemMarker.kt
package galenscovell.hinterstar.generation.sector

import com.badlogic.gdx.graphics.g2d.*
import com.badlogic.gdx.scenes.scene2d.*
import com.badlogic.gdx.scenes.scene2d.utils.ActorGestureListener
import galenscovell.hinterstar.util.*


class SystemMarker(x: Int, y: Int) : Actor() {
    val sx: Float = x * Constants.SYSTEMMARKER_SIZE
    val sy: Float = y * Constants.SYSTEMMARKER_SIZE

    private var frames: Int = 60
    private var markerType: MarkerType = MarkerType.EMPTY
    private var glowing: Boolean = true

    lateinit private var system: System
    lateinit private var sprite: Sprite



    /********************
     *     Getters     *
     ********************/
    fun getThisSystemMarker(): SystemMarker {
        return this
    }

    fun getSystem(): System {
        return system
    }

    fun isEmpty(): Boolean {
        return markerType == MarkerType.EMPTY
    }

    fun isCurrent(): Boolean {
        return markerType == MarkerType.CURRENT
    }

    fun isExplored(): Boolean {
        return markerType == MarkerType.EXPLORED
    }

    fun isUnexplored(): Boolean {
        return markerType == MarkerType.UNEXPLORED
    }



    /********************
     *     Setters     *
     ********************/
    fun setSystem(s: System): Unit {
        system = s
        this.addListener(object : ActorGestureListener() {
            override fun touchDown(event: InputEvent, x: Float, y: Float, pointer: Int, button: Int): Unit {
                SystemOperations.setSelection(s)
            }
        })
    }

    fun becomeEmpty(): Unit {
        markerType = MarkerType.EMPTY
    }

    fun becomeCurrent(): Unit {
        Resources.currentMarker.setTarget(
                sx + Constants.SYSTEM_MARKER_X - Constants.SYSTEMMARKER_SIZE,
                Constants.EXACT_Y - sy + Constants.SYSTEM_MARKER_Y - Constants.SYSTEMMARKER_SIZE
        )
        markerType = MarkerType.CURRENT
    }

    fun becomeExplored(): Unit {
        markerType = MarkerType.EXPLORED
    }

    fun becomeUnexplored(): Unit {
        sprite = Resources.spTest0
        markerType = MarkerType.UNEXPLORED
    }



    /**********************
     *     Rendering     *
     **********************/
    override fun draw(batch: Batch, parentAlpha: Float): Unit {
        if (!isEmpty()) {
            glow(batch)
            batch.draw(
                    sprite,
                    sx + Constants.SYSTEM_MARKER_X,
                    Constants.EXACT_Y - sy + Constants.SYSTEM_MARKER_Y,
                    Constants.SYSTEMMARKER_SIZE,
                    Constants.SYSTEMMARKER_SIZE
            )
            if (isCurrent()) {
                Resources.currentMarker.render(batch)
            }
        }
    }

    private fun glow(batch: Batch): Unit {
        if (glowing) {
            frames += 1
        } else {
            frames -= 2
        }

        if (frames == 120) {
            glowing = false
        } else if (frames == 40) {
            glowing = true
        }

        val frameAlpha: Float = frames / 120.0f
        if (isExplored()) {
            batch.setColor(0.2f, 0.2f, 0.75f, frameAlpha)
        } else {
            batch.setColor(0.2f, 0.75f, 0.2f, frameAlpha)
        }

        batch.draw(
                Resources.mapGlow,
                sx + Constants.SYSTEM_MARKER_X - Constants.SYSTEMMARKER_SIZE,
                Constants.EXACT_Y - sy + Constants.SYSTEM_MARKER_Y - Constants.SYSTEMMARKER_SIZE,
                Constants.SYSTEMMARKER_SIZE * 3,
                Constants.SYSTEMMARKER_SIZE * 3
        )
        batch.setColor(1f, 1f, 1f, 1f)
    }

    fun debugDraw(): String {
        when (markerType) {
            MarkerType.EMPTY -> return "~"
            MarkerType.UNEXPLORED -> return "U"
            MarkerType.EXPLORED -> return "E"
            MarkerType.CURRENT -> return "C"
            else -> return "!"
        }
    }
}
